pkg/api/apikey.go

```diff
@@ -76,11 +76,22 @@ func (hs *HTTPServer) AddAPIKey(c *models.ReqContext, cmd models.AddApiKeyComman
 	var serviceAccount *models.User = &models.User{Id: -1}
 	if hs.Cfg.FeatureToggles["service-accounts"] {
 		if cmd.CreateNewServiceAccount {
-			serviceAccount, err = hs.AccessControl.CloneUserToServiceAccount(c.Req.Context(), c.SignedInUser)
+			serviceAccount, err = hs.SQLStore.CloneUserToServiceAccount(c.Req.Context(), c.SignedInUser)
 			if err != nil {
 				return response.Error(500, "Unable to clone user to service account", err)
 			}
 			cmd.ServiceAccountId = serviceAccount.Id
+		} else {
+			//Check if user and service account are in the same org
+			query := models.GetUserByIdQuery{Id: serviceAccount.Id}
+			err = bus.DispatchCtx(c.Req.Context(), &query)
+			if err != nil {
+				return response.Error(500, "Unable to clone user to service account", err)
+			}
+			serviceAccountDetails := query.Result
+			if serviceAccountDetails.OrgId != c.OrgId || serviceAccountDetails.OrgId != cmd.OrgId {
+				return response.Error(403, "Target service is not in the same organisation as requesting user or api key", err)
+			}
```

Comment:
I got confused when reading this again.
if serviceaccounts - feature, if NOT cmd.Serviceaccount:

we will not have the serviceaccount in this closure/scope. what is this really for?

Reply:
I'll add comments to make it more clear